Bitstream tracking for replacements and copies

ConfusionOrb221 opened the pull request:
## References
- Fixes https://github.com/DSpace/DSpace/issues/10504
- Angular PR: https://github.com/DSpace/dspace-angular/pull/4368

## Description
This PR introduces REST endpoints to replace existing Item Bitstreams
- From the edit Item page
- From the submission form, when submitting a new version of an existing Item

It also introduces special metadata fields to denote links between related Bitstreams:
- Item versioning
    - `dspace.bitstream.isCopyOf`
        - Links to the original Bitstream from the previous version
    - `dspace.bitstream.hasCopies`
        - Links to the cloned Bitstream from the subsequent version(s)
        - Because DSpace versioning allows Items to be "forked", this field can contain multiple values
- Bitstream replacement
    - `dspace.bitstream.isReplacementOf`
        - Links to the original Bitstream that was replaced
    - `dspace.bitstream.isReplacedBy`
        - Links to the new Bitstream that replaced the current one
- All of the above fields store Bitstream UUIDs, and should not be used outside `BitstreamLinkingService`
    - _Note: depending on the discussion surrounding the CRIS merger we may want to store UUIDs as authority values instead_
###### Future work
In future PRs, the `dspace.bitstream.*` metadata fields can be used to
- Show links to original & replacement versions on the edit Bitstream page
- Redirect `GET /content` requests to the replacement Bitstream

![image](https://github.com/user-attachments/assets/52dd2249-7e4c-484a-8da0-31e4bc080b2b)


## Instructions for Reviewers
- Adds `BitstreamLinkingService` which defined the logic for linking bitstreams to each other
- Adjusts `BitstreamService` to call upon the `linkingService` when bitstreams are cloned
- Adds replacement functionality to `BitstreamService`
- Adds our new DSpace fields to `dspace-types.xml`
- Extended `BitstreamServiceImpl` with a new method replace to replace bitstreams.
- Created `BitstreamReplaceController` that supports a new REST endpoint to replace bitstreams.
- Changed the return type of `UploadableStep.upload` from `ErrorRest` to `Pair<Bitstream, ErrorRest>` to allow SubmissionService to retrieve the new uploaded file.
- Updated `SubmissionService` at `uploadFileToInprogressSubmission` to replace uploaded bitstreams when the replaceFile parameter is passed.

How to test or review the PR:
- Refer to Angular PR

## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/BitstreamReplaceController.java:
```diff
@@ -0,0 +1,100 @@
+
+/**
+ * Controller to replace bitstreams in a bundle.
+ * Endpoint: /api/core/bitstreams/{uuid}/replace
```
@ConfusionOrb221 : This new endpoint is undocumented, but it appears to disobey our REST API rules for ["Naming of Endpoints"](https://github.com/DSpace/RestContract/tree/main?tab=readme-ov-file#on-the-naming-of-endpoints) where we say that all "endpoint names MUST use plural nouns instead of verbs" 

So, we cannot use the word `/replace` here as that's a verb.

Couldn't we change this to be a `PUT /api/core/bitstreams/{uuid}/content`?  By definition `PUT` is a replacement, and this endpoint would be replacing just the `/content` while leaving the metadata in tact.

steph-ieffam commented on dspace-api/src/main/java/org/dspace/content/service/BitstreamService.java:
```diff
@@ -85,6 +85,18 @@ public interface BitstreamService extends DSpaceObjectService<Bitstream>, DSpace
         throws IOException, SQLException, AuthorizeException;
 
+    /**
+     * Replaces {@code oldBitstream} with Replaces {@code oldBitstream}.
```
there are typos in these lines

nwoodward replied:
Still needs to be changed. Please remove the second "Replaces". And this line and the line after it refer to `oldBitstream` twice.

nwoodward replied:
This line and the line below it still need to be fixed.

ybnd replied:
Done, also clarified that the old (& derived) Bitstreams get deleted

steph-ieffam reviewed (changes requested):
@ConfusionOrb221 I have added some comments but rather than actual requested changes I think a bit of discussion is needed about how to call the new endpoint and how to better use and retrieve the added metadata. 
I'm also a bit worried/confused about the naming of the methods which I think is a bit misleading

ybnd commented on dspace-api/src/main/java/org/dspace/content/BitstreamLinkingServiceImpl.java:
```diff
@@ -0,0 +1,180 @@
+    @Override
+    public void registerReplacementBitstream(Context context, Bitstream oldCopy,
+                                             Bitstream replacementCopy) throws SQLException, AuthorizeException {
+        bitstreamService.addMetadata(context, oldCopy, DSPACE, BITSTREAM,
```
> There seems to be no way to get this data from the UI without moving to the Administrative metadata tab
> ... also pretty hard to retrieve 

That's correct at the moment, this is not covered by these first two PRs.
Further UI would be a next step; if you/@tdonohue think it would make it a more complete package we can include it already.

> I think what could be an improvement would be storing the name of the bitstream in the Value field and the UUID in the Authority

We can do this, but maybe it would have to depend on the CRIS merger discussion.

> If the bitstream is being marked as deleted the bitstream data is going to be deleted as soon as the cleanup script is ran so does it really makes sense to keep track of this information using metadata?

We'd have to depend on those rows staying in the database ~ `./dspace cleanup --leave`

ybnd commented on dspace-api/src/main/java/org/dspace/content/BitstreamLinkingServiceImpl.java:
```diff
@@ -0,0 +1,180 @@
+    public static final String HAS_COPIES = "hasCopies";
+    public static final String IS_COPY_OF = "isCopyOf";
+    public static final String IS_REPLACED_BY = "isReplacedBy";
+    public static final String IS_REPLACEMENT_OF = "isReplacementOf";
```
@tdonohue we replaced the original PR description, and I wrote out a more detailed step-by-step on the wiki to hopefully clarify further: https://wiki.lyrasis.org/display/DSDOC10x/Bitstream+versioning+and+replacement

tdonohue replied:
@ybnd : Thanks for adding more documentation around this feature. Apologies if I'm being "dense", but I'm still a bit confused by the diagrams in the examples you've added.  I've tried to explain my confusion using [inline comments on that wiki page](https://wiki.lyrasis.org/display/DSDOC10x/Bitstream+versioning+and+replacement).  Could you take a look at my comments and clarify the behavior (or enhance the docs further to describe what is going on)?

ybnd commented:
@tdonohue as you mentioned in https://wiki.lyrasis.org/display/DSDOC10x/Bitstream+versioning+and+replacement

> We may need a way to avoid this scenario, and default  to the --leave  behavior whenever Item Versioning is enabled.

Now implemented in [cddd860](https://github.com/DSpace/DSpace/pull/10554/commits/cddd860fe1a9a64a528a6c5b097199cdfdf8cda4)

nwoodward commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/BitstreamReplaceController.java:
```diff
@@ -0,0 +1,100 @@
+        Context context = ContextUtil.obtainContext(request);
+        Bitstream bitstream = bitstreamService.find(context, uuid);
+        if (bitstream == null || bitstream.isDeleted()) {
+            throw new ResourceNotFoundException("The bitstream with ID:" + uuid + "could not be found");
```
Add a space before "could" to separate it from the UUID.

nwoodward replied:
Can be fixed later.

nwoodward commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java:
```diff
@@ -181,20 +183,20 @@ public Class<WorkspaceItemRest> getDomainClass() {
         List<ErrorRest> errors = submissionService.uploadFileToInprogressSubmission(context, request, wsi, source,
                 file);
+        context.commit();
+        context.reloadEntity(source);
```
This feedback is still outstanding. @ybnd as you say, the errors added to `wsi` are just for the UI, but it's unclear if there was a **need** to move up the `commit()`. My guess is it was because it may generate additional errors that would rise up when `wsi = converter.toRest(source, utils.obtainProjection());` is called, but what could those errors be?

nwoodward replied:
Is there an explanation for why the `commit()` had to be moved up?

ybnd replied:
I dug this up from my notes (almost exactly a year ago, time flies)

> Looks like the **missing access conditions** were a REST issue after all: they’re replaced correctly but not taken into account once `converter.toRest` is called.

Apparently the fix was not working, because we weren't actually using the reloaded `source` object (!)

Confirmed that the access conditions are indeed missing from the POST response if we revert this change.
- Note that the UI impact is not all that great: we end up sending a GET request right after where the access conditions _do_ come through
- IMO it's still worth keeping the fix so we can eliminate that second request later

I've updated one of the tests to check that the access condition change actually comes through and added an inline comment to clarify.

tdonohue commented on dspace/config/modules/versioning.cfg:
```diff
@@ -5,7 +5,11 @@
+
+# The property replace-bitstream.enabled is used to enable/disable bitstream replacement in DSpace,
+# the default value is false if unset
+# replace-bitstream.enabled = false 
```
Tiny thing, could we consider renaming this to be `versioning.replace-bitstream.enabled`?  All the other configs in this file are prefixed with `versioning.*`.  So, we should consider renaming this for consistency.

NOTE: Obviously, if we rename this it also needs to be renamed in the frontend PR since this config is read by the frontend as well.

ybnd replied:
Bitstream replacements can work with versioning disabled, so maybe it would be more appropriate to move this property to another configuration file?

tdonohue replied:
@ybnd : Ok, that makes more sense. I was not aware that this feature can be used without versioning.

If you want to keep this setting as-is in the `versioning.cfg` file, I'm OK with it.  The only other place would could move it is the `dspace.cfg`. 

However, I think the description of this setting is insufficient.  We need to update the comments above this setting to describe how it is used, what it does and clarify that it can enhance Item versioning but also can be used standalone (without versioning).

So, it should clarify that this single setting does these things:
1. Enables the replace button next to bitstreams in the User Interface
2. Enables linking/tracking of replacement bitstreams with the original bitstream
3. Helps to enhance Item versioning by making it clear when Bitstream replacements occurred, but it's a standalone feature that can be used whether or not Item versioning is enabled.

tdonohue commented on dspace-api/src/main/java/org/dspace/storage/bitstore/Cleanup.java:
```diff
@@ -53,37 +65,54 @@ public static void main(String[] argv) {
+            boolean versioning = new DSpace().getConfigurationService().getBooleanProperty("versioning.enabled", true);
+            boolean replaceBitstream = new DSpace().getConfigurationService()
+                    .getBooleanProperty("replace-bitstream.enabled", false);
+            boolean defaultLeave = versioning || replaceBitstream;
```
Shouldn't this be `versioning && replaceBitstream`?

It seems like you should never have `replace-bitstream.enabled=true` while `versioning.enabled=false` because replacing a bitstream requires versioning to be enabled.  Or am I misunderstanding things here?

Overall, because `versioning.enabled=true` by default, then `defaultLeave` is going to ALWAYS be `true` unless you specifically turn off versioning.  That would change the default behavior to always leaving the bitstream data in the database.

ybnd replied:
It should be possible to only enable `replace-bitstream.enabled=true` by itself.
In that case you can only replace from the "Bitstream edit" page.

Either way, as soon as either feature is enabled

> change the default behavior to always leaving the bitstream data in the database

is the intent, in order to maintain metadata links (discussed above: https://github.com/DSpace/DSpace/pull/10554#discussion_r2602995244)

tdonohue replied:
This was clarified above. I now understand this setting can be used without versioning. That wasn't clear previously.

tdonohue reviewed (commented):
@ybnd and @ConfusionOrb221 :  I've reviewed and tested this today alongside the frontend PR.  I've noted the main bugs I've encountered in [my frontend PR review](https://github.com/DSpace/dspace-angular/pull/4368#pullrequestreview-4076991653), because the bugs all appear to be on the frontend.

In terms of the backend, I only have a few minor issues that I've noted inline below.  The rest of the code looks good.

I also did verify that this backend PR is linking related BItstreams via the Authority Framework (as was requested). So, I verified that I'm seeing behavior like this:
* **If Bitstream-1 is replaced by Bistream-2** 
    * Bitstream-1 will have a `dspace.bitstream.isReplacedBy` metadata field that has a `value` equal to Bitstream-2's name and an `authority` equal to Bistream-2's UUID.  
    * Conversely,  Bitstream-2 will have a `dspace.bitstream.isReplacementOf` metadata field that has a `value` equal to Bitstream-1's name and an `authority` equal to Bistream-1's UUID. 
* **Similarly, if you create a new Version of an Item that has a single Bitstream**
    * The original Bitstream will be copied to a Copy-Bitstream which has a `dspace.bitstream.isCopyOf` metadata field that has a `value` equal to the original Bitstream's name  and an `authority` equal to the original bitstream's UUID.
    * The opposite relationship also exists via `dspace.bitstream.hasCopies`

Overall, this backend PR looks good to me. But, I'd appreciate it if you could look at my two comments below.

atarix83 reviewed (changes requested):
Thanks @ConfusionOrb221 for this PR!

I tested it together with the corresponding frontend PR, and I have a few points/concerns:

- Derived bitstream handling: I share the same concern raised by @nwoodward regarding derived bitstreams. If we cannot remove the derived bitstream associated with the replaced original bitstream, we should at least display a warning on the replace page to inform the user about the inconsistency this operation will create.


- dspace.bitstream.isReplacedBy metadata: Perhaps I’m misunderstanding the expected behavior, but I’m unable to verify whether this metadata is correctly added. After replacing the original bitstream, I’m unable to retrieve it anymore (I get a 404). Should the original bitstream still exist after replacement, or is it expected to be removed?


- Configuration options: Concerning the enable/disable flags for this feature, would it be possible to separate the behavior for replacing a bitstream during versioning from the behavior for replacing it via the Admin UI? Having more granular control over where replacement is allowed would be very helpful.

ybnd commented:
Thanks again for the detailed feedback everyone!
We aimed to get everything addressed at once to avoid spamming the PRs again. In the most recent push you'll see
- `THUMBNAIL` and `TEXT` Bitstreams are now cleared after replacing a Bitstream, so new one will be picked up in the next `filter-media` run
    - Implemented in a way that follows the `FormatFilter` plugin system, instead of configuring derived Bundles separately
    - In order to make this possible we exposed the existing logic to list derived Bitstreams from `MediaFilterService`
- Fixed the Bitstream format not getting determined on new Bitstreams after replacement
- Added two independently configurable authorization features:
    - `canReplaceBitstreamAdmin` for the "Administer" pages (property: `replace-bitstream.ui.admin`)
    - `canReplaceBitstreamSubmitter` for the "Edit" page (property: `replace-bitstream.ui.submitter`)
    - Both of these respect the global `replace-bitstream.enabled` property

@nwoodward We weren't able to definitively replicate the 
> `The bitstream with ID:446bf66a-4d3a-489d-9fb8-8d98b2e18fa4 could not be found`
error, perhaps something was going wrong on the Angular side
- For instance, the request may have been sent with a UUID that had already been replaced
- We've addressed a few cache invalidation issues in the Angular side, I think it's worth re-testing to see if that was the cause

@atarix83
> Should the original bitstream still exist after replacement, or is it expected to be removed?

Those Bitstreams are indeed expected to be removed, so currently those metadata can only be accessed within the DSpace backend or the database.

ybnd commented:
@tdonohue looks like the Docker job is failing on the new `authority.required.*` configurations [here](https://github.com/DSpace/DSpace/pull/10554/changes#diff-f96f33c453485bfba1e274ecb8881e997a348dc1ad7b5cf5c8c9246b1740aa4bR550-R553). This was done to enforce that authority values are present, since we now read Bitstream UUIDs from there.

If I understand correctly, those jobs use the [Entities demo DB](https://github.com/DSpace-Labs/AIP-Files/releases/download/demo-entities-data/dspace7-entities-data.sql), where the `RegistryUpdater` migration is already done, so new fields are not added to the registry anymore.

We have a few options on how to go forward, but I'm not sure what would be best:
- Undo the `authority.required.*` configuration changes
  - We'd lose our sanity check
- Update the Docker job/setup/DB to ensure those fields are loaded into the registry
  - This would solve it in CI, but the same problem would reappear for anyone upgrading a DSpace instance with an existing database
- Add a PSQL migration to create those fields manually

atarix83 commented:
> @tdonohue looks like the Docker job is failing on the new `authority.required.*` configurations [here](https://github.com/DSpace/DSpace/pull/10554/changes#diff-f96f33c453485bfba1e274ecb8881e997a348dc1ad7b5cf5c8c9246b1740aa4bR550-R553). This was done to enforce that authority values are present, since we now read Bitstream UUIDs from there.
> 
> If I understand correctly, those jobs use the [Entities demo DB](https://github.com/DSpace-Labs/AIP-Files/releases/download/demo-entities-data/dspace7-entities-data.sql), where the `RegistryUpdater` migration is already done, so new fields are not added to the registry anymore.
> 
> We have a few options on how to go forward, but I'm not sure what would be best:
> 
>     * Undo the `authority.required.*` configuration changes
>       
>       * We'd lose our sanity check
> 
>     * Update the Docker job/setup/DB to ensure those fields are loaded into the registry
>       
>       * This would solve it in CI, but the same problem would reappear for anyone upgrading a DSpace instance with an existing database
> 
>     * Add a PSQL migration to create those fields manually

@ybnd  @tdonohue 
I reckon we should add the registry loader execution to the pipeline. For users migrating from a previous version, this should not be an issue, as the Flyway migration will also trigger the metadata registry update.

tdonohue commented:
@ybnd : The `dspace-deploy` Docker script is **not** using the Entities DB data.  Instead, it's starting with a fresh, empty database and loading in some test AIPs using [`cli.ingest.yml`](https://github.com/DSpace/DSpace/blob/main/dspace/src/main/docker-compose/cli.ingest.yml)

The odd thing about the error in this PR is that the error seems to be triggered **before** `dspace-deploy` even loads AIP data. The error is occurring when `./dspace create-administrator` is being run here: 
https://github.com/DSpace/DSpace/blob/main/.github/workflows/docker.yml#L190

At the point this error occurs this means the database should be empty and has been initialized by our main `docker-compose.yml` file [which runs `./dspace database migrate`](https://github.com/DSpace/DSpace/blob/main/docker-compose.yml#L65) and that command should result in all Metadata fields being initialized (as that will have run database migrations which then triggers `RegistryUpdater`).

I'm also very confused as to why the `./dspace create-administrator` command even triggers the `authority.required.*` configurations to load?  I'm not understanding why creating an Admin account results in these settings being read?  That seems like odd behavior to me because we don't need bitstream tracking information when creating an Administrator account.

I have to admit, I'm confused by this error.  I'm not yet convinced it's a flaw in the Docker job or setup.  It seems like this PR is somehow affecting the behavior of the `create-administrator` script, which it likely should **not** be doing

**UPDATE:**  Oh, I think I see.  The `authority.required.*` settings are loaded when DSpace initializes because `MetadataAuthorityServiceImpl` is [greedily loading all this information in `init()`](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/content/authority/MetadataAuthorityServiceImpl.java#L119).  (That's not really the best behavior, but it's not necessarily the job of this PR to fix that issue.)

**UPDATE 2:** It's **possible** that updating our `docker-compose.yml` to use `dspace database migrate force` will fix the issue. See [this line](https://github.com/DSpace/DSpace/blob/main/docker-compose.yml#L65).  The "force" command will force the various callbacks (like `RegistryUpdater`) to run **even if there are no database migrations**.  But, I'm still surprised that `RegistryUpdater` is not already triggered because `docker-deploy` should start from an empty database and therefore all migrations should run to create the database tables.

tdonohue commented:
@ybnd : I  have a new theory as to why the automated tests are failing here.  I think it **might be a race condition**... essentially the `./dspace create-administrator` command is triggering **before** the database has been fully initialized (i.e. the migrations have not yet completed or the metadata fields have not yet been loaded).

I notice when I test this PR locally with a branch new database, everything works fine.  But, it does take a while to load all the metadata fields into a fresh database.  I suspect, by chance, these extra metadata fields in this PR made it finally take just a bit too long.

Therefore, I've created a small PR which should fix a potential "race condition" by updating the `docker-deploy` job to wait for the REST API to become responsive before it runs `./dspace create-administrator`.  See #12320

Could you try cherry-picking the commit in #12320 over to this PR to see if it fixes the problem?   I'd also recommend rebasing this PR on the latest `main` and squash the 56 commits (if that's possible). You don't have to turn this whole PR into one commit, but 56 is a *very large number of commits*...which means I'll likely do a "Squash & merge" on this PR later on (to turn it into one commit) unless it's cleaned up.

Finally, just wanted to add that I retested this backend with the frontend today!  Everything is working well except for a small issue that I'm still having on the frontend (see that review for details)

ybnd committed:
Bitstream replacement & metadata links

Co-Authored-By: Nathan Buckingham <nate1898@gmail.com>
Co-Authored-By: Joran De Braekeleer <joran.debraekeleer@atmire.com>
Co-Authored-By: Jens Vannerum <jens.vannerum@atmire.com>

ybnd committed:
Merge remote-tracking branch 'origin/main' into feature-bitstream-linking-7.6

ybnd committed:
Adjust to DSpace 10, support /edititems endpoint

Co-Authored-By: Jens Vannerum <jens.vannerum@atmire.com>

nwoodward commented:
@ynbd Thanks for all of the updates. Overall, it's working better and many of the issues raised above have been resolved. I confirmed that derivatives are replaced by `filter-media` after a new version of a bitstream is uploaded.

One thing I noticed while testing that functionality is when I click on Replace a bitstream, I upload a file, and I click Save then I land on the bitstream edit page with a message like this at the top.
<img width="1308" height="64" alt="Screenshot 2026-04-29 at 10 15 17 AM" src="https://github.com/user-attachments/assets/7c85915f-2039-46f2-b8d7-f1f1b97dfa59" />

I looked in the database, and the old bitstream has already been replaced at this point. If I scroll down on the bitstream edit form and click the Cancel button it's too late. Whether I click on Cancel or Save I return to the Administer page, and the new bitstream is there.

It's a small thing, and I'm not sure if there's a solution, because I think it's wise for the bitstream replacement functionality to look and act much like the upload and edit bitstream pages. All it might require is a message on the replace bitstream page that says once the user clicks Save this action can't be undone.

ybnd committed:
Change submission upload replaceFile parameter to UUID

This synchronizes the /workspaceitems with the /edititems, where we don't have a clear file order

ybnd commented:
@tdonohue 
As mentioned in https://github.com/DSpace/dspace-angular/pull/4368#issuecomment-4352659822: I've synced how the `replaceFile` parameter is treated between the "edit" and "submission" endpoints.

tdonohue commented on dspace-server-webapp/javac.20260430_102156.args:
```diff
@@ -0,0 +1 @@
+@/private/var/folders/tl/xxzvtv5d2c150_sb044jrqw40000gn/T/org.codehaus.plexus.compiler.javac.JavacCompiler10740251155750947781arguments
```
Please remove this accidentally added file.  Thanks!

ybnd replied:
Done ✅

tdonohue reviewed (changes requested):
@ybnd and @ConfusionOrb221 :  I retested this alongside the frontend PR ([see that review](https://github.com/DSpace/dspace-angular/pull/4368#pullrequestreview-4208463288)).  I'm basically a +1, but it looks like a `javac.*.args` file has been accidentally added in the last commit.  Once that is removed, then I'm a +1 for this PR as well.

Thanks!

tdonohue reviewed (approved):
👍 Thanks @ybnd and @ConfusionOrb221 !  This looks good to me now.

nwoodward commented:
One idea for another potential follow-up pull request... right now I can get into odd situations with bitstream versioning of thumbnails. For example, if I add a bitstream in the ORIGINAL bundle and run `filter-media` to generate a thumbnail bitstream in the THUMBNAIL bundle then I am able to replace that thumbnail bitstream with a new bitstream. And if I leave the new file name toggle on in the form then I end up with two generated thumbnail bitstreams, one of which is for a bitstream that isn't in the ORIGINAL bundle anymore. Even if I run `filter-media -f` I'm left with the two generated thumbnail bitstreams.

It would be nice if along with all the other bitstream versioning configuration in `config/modules/versioning.cfg` there was a parameter for a list of bundles where this functionality is enabled. In our case, I would only want it enabled for the ORIGINAL bundle. Just a thought.

ybnd committed:
Fix documentation on BitstreamService.replace

ybnd committed:
Ensure Bitstream access conditions appear after replacing

tdonohue commented:
As this is at +2, I'm merging this for inclusion in 10.0.  Any further work can occur in follow-up PRs.

@ybnd and @ConfusionOrb221 : I am flagging this as `needs documentation` because the new settings and feature need to be documented in our [DSpace 10.0 docs](https://wiki.lyrasis.org/spaces/DSDOC10x/pages/408944697/DSpace+10.x+Documentation).  Please let me know once you've documented this feature and we can remove that flag.
